File: app/src/main/java/com/nawin/androidmvparchitecture/data/local/LocalRepo.java
package com.nawin.androidmvparchitecture.data.local;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import android.util.Log;

import com.google.gson.Gson;
import com.nawin.androidmvparchitecture.data.model.UserInfo;

/**
 * Created by brainovation on 6/13/17.
 */

public class LocalRepo {

    private static final String USER_INFO = "_user_info";
    private final SharedPreferences sharedPreferences;
    private final Gson gson;
    private static UserInfo cachedUserInfo;
    private static LocalRepo localRepo;

    public static LocalRepo getInstance(Context context) {
        if (localRepo == null) {
            localRepo = new LocalRepo(context, new Gson());
        }
        return localRepo;
    }

    public LocalRepo(Context context, Gson gson) {
        this.sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        this.gson = gson;
    }

    public void setUserInfo(@NonNull UserInfo userInfo) {
        sharedPreferences.edit().putString(USER_INFO, gson.toJson(userInfo)).apply();
        cachedUserInfo = userInfo;
    }

    public UserInfo getUserInfo() {
        if (cachedUserInfo == null) {
            String userInfoS = sharedPreferences.getString(USER_INFO, null);
            if (!TextUtils.isEmpty(userInfoS)) {
                cachedUserInfo = gson.fromJson(userInfoS, UserInfo.class);
            }
        }
        return cachedUserInfo;
    }
}

File: app/src/main/java/com/nawin/androidmvparchitecture/data/local/LocalRepo.kt
package com.nawin.androidmvparchitecture.data.local

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.text.TextUtils
import com.google.gson.Gson
import com.nawin.androidmvparchitecture.data.model.UserInfo
import javax.inject.Inject

/**
 * Created on 11/29/17.
 */
class LocalRepo @Inject constructor(context: Context, private val gson: Gson) {

    private val USER_INFO = "_user_info"
    private val sharedPreferences: SharedPreferences =
            PreferenceManager.getDefaultSharedPreferences(context)
    private var cachedUserInfo: UserInfo? = null

    fun setUserInfo(userInfo: UserInfo) {
        sharedPreferences.edit().putString(USER_INFO, gson.toJson(userInfo)).apply()
        cachedUserInfo = userInfo
    }

    fun getUserInfo(): UserInfo? {
        if (cachedUserInfo == null) {
            val userInfoS = sharedPreferences.getString(USER_INFO, null)
            if (!TextUtils.isEmpty(userInfoS)) {
                cachedUserInfo = gson.fromJson(userInfoS, UserInfo::class.java)
            }
        }
        return cachedUserInfo
    }

    fun logout() {
        this.cachedUserInfo = null
        sharedPreferences.edit().clear().apply()
    }
}